text/0017-shared-config.md

```diff
@@ -0,0 +1,120 @@
+# Configuration
+- Start Date: 2019/11/11
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/16
+
+## Definitions
+
+**Config files** allow implementers to customize the behavior of microfrontends
+by providing static files. Common formats include JSON, YAML, and HCL.
+
+A **config schema** is the set of expectations that a microfrontend has of
+implementation-provided config files. It says how the config file must be 
+structured, what keys it must have, and constrains what types of values are allowed.
+
+A **config library** is a code module that allows developers to define config 
+schemas, and loads config files in accordance with that schema. It may also 
+provide other tooling related to configuration.
+
+**Shared config** is configuration values that are shared between distinct modules.
+
+
+## Decision
+
+Config elements that will be shared between multiple modules should be added to a `common`
+config tree. The values provided to this config tree would be made available to any module
```

Comment:
> The values provided to this config tree would be made available to any module that requested them in its schema.

This means that the `getConfig('@openmrs/esm-login').then(config => {})` would return the common config?

I'm nervous about that. If both the common and module-specific configs have the same key, there will be the issue of merging. I'd prefer this more explicit workflow instead:

```js
const moduleConfig = useConfig()
const commonConfig = useConfig('common')
```

Reply:
What I’m proposing is that elements of common would need to be explicitly “included” in the config schema. See the example—
`myFoo: common.foo`

Reply:
To clarify: `getConfig('@openmrs/esm-login').then(config => {})` would not return the common config. Given a schema like
```
{
  baz: { default: "bazzz" },
  myFoo: common.foo
}
```
it would return a config like
```
{
  baz: "customBaz",
  myFoo: "customFoo"
}
```

i.e., the common config elements have module-specified insertion points into the module's configuration.

Reply:
When defining the schema, where does `common` come from? It looks like a variable, not a string. Does it need to be imported somewhere?

Reply:
I'm open to recommendations, but I suppose my first inclination would be to make it importable from `@openmrs/esm-module-config`, like
```javascript
import { defineConfigSchema, commonConfig } from "@openmrs/esm-module-config"

defineConfigSchema("my-module", {
  baz: { default: "bazzz" },
  myFoo: commonConfig.foo
})
```

Also very open to suggestions on the name. `common`? `commonConfig`? `commonSchemas`? `commonConfigSchemaComponents`? Not sure what the right level of verbosity is.

Another option might be
```javascript
import { defineConfigSchema } from "@openmrs/esm-module-config"
import { nameConfig } from "@openmrs/esm-module-config/common"

defineConfigSchema("my-module", {
  baz: { default: "bazzz" },
  myFoo: nameConfig
})
```

Reply:
I see. I like the former.